tree_test.go

```diff
@@ -169,6 +170,7 @@ func TestTreeWildcard(t *testing.T) {
 		{"/search/", false, "/search/", nil},
 		{"/search/someth!ng+in+ünìcodé", false, "/search/:query", Params{Param{Key: "query", Value: "someth!ng+in+ünìcodé"}}},
 		{"/search/someth!ng+in+ünìcodé/", true, "", Params{Param{Key: "query", Value: "someth!ng+in+ünìcodé"}}},
+		{"/search/gin", false, "/search/:query", Params{Param{"query", "gin"}}},
```

Comment:
I think this is more complicated than the proposed fix above.

Try adding some more routes to cover some bases like these.  Because of the tree structure, this doesn't look like it will incorrectly handle when these routes exist

-  `/search/gin-gonic`
- `/search/google`
- `/search/:param`

If you send in `/search/gaga` which route is hit? The last one?

I think we need a lot of tests, which I failed to do on the original PR.

I'm sorry that I haven't had much time to tackle this, but I believe that the best answer involves some lightweight backtracking that can be done cheaply.